Add --exclude option to rpc-tests.py

jnewbery opened the pull request:
Sometimes it's useful to be able to run all test *except* some specified tests, either because those tests take a long time to execute or because you know they're broken.

This commit adds an additional `--exclude` or `-x` option to rpc-tests to exclude a comma-separated-list of tests. Usage is:

`rpc-tests.py --exclude=test1,test2,test3`
or
`rpc-tests.py --exclude test1,test2,test3`
or
`rpc-tests.py -x=test1,test2,test3`
or
`rpc-tests.py -x test1,test2,test3`

test names are given without the `.py` suffix.

This can be combined with the extended option to run the entire suite of tests, except for a single test, eg pruning.py which takes a long time:

`rpc-tests.py --extended -x pruning`

Note for reviewers: this requires and builds on top of https://github.com/bitcoin/bitcoin/pull/9657 . Please review only the last commit in this PR (921f6e1ac9b32cb80a83d97054b4cfdbbc290110)

[EDIT: edited commit hash]

jnewbery committed:
Add exclude option to rpc-tests.py

jnewbery commented:
@morcos

jtimon commented on qa/pull-tester/rpc-tests.py:
```diff
@@ -194,6 +195,12 @@ def main():
 
+    # Remove the test cases that the user has explicitly asked to exclude.
+    if args.exclude:
+        [ test_list.remove(exclude_test + ".py")
```
why use [ ] here?

I think this would be more clear:
```
    if args.exclude:
        for exclude_test in args.exclude.split(','):
            if exclude_test + ".py" in test_list:
                test_list.remove(exclude_test + ".py")
```
And it's the same number of lines.

jnewbery replied:
ha. If I write for loops I get review comments saying I should use list comprehensions, and if I write list comprehensions I get review comments saying I should use for loops :)

In this case you're probably right that the for loop is clearer.

jtimon commented:
utACK f34e664 (although I left a style nit).
Yay, ```--extended --exclude=pruning```!!

MarcoFalke commented:
Concept ACK!

jnewbery commented:
I've changed the list comprehension to a for loop as suggested by @jtimon . I've also moved the `if len(test_list) == 0` block lower down to catch the (unlikely) failure case where the user specifies and excludes the same test.

jnewbery commented:
rebased

jtimon commented on qa/pull-tester/rpc-tests.py:
```diff
@@ -198,6 +192,17 @@ def main():
+    # Remove the test cases that the user has explicitly asked to exclude.
+    if args.exclude:
+        for exclude_test in args.exclude.split(','):
+            if exclude_test + ".py" in test_list:
```
Don't you need the following?
```
            if exclude_test in test_list:
                test_list.remove(exclude_test)
            if exclude_test + ".py" in test_list:
                test_list.remove(exclude_test + ".py")
```

Otherwise --exclude won't work if you specify the test with including .py explicitly.
Oh, sorry, the --exclude documentation explains it, never mind.

jtimon commented:
re utACK c578408

MarcoFalke commented:
utACK c5784085e9031ea1ad32a2aee897bd57a0482d24
